File: Java/20220905_Java_Exceções/src/Conta.java
//Declaração de classe abstrata
public abstract class Conta {
    //Declaração de atributos com seus respectivos tipos
    private static int total;
    protected double saldo;//Transformado em protected para uso nas classes herdeiras
    private int agencia;
    private int numero;

    //Declaração de construtor
    public Conta(int agencia, int numero){
        this.agencia = agencia;
        this.numero = numero;
        //
        total++;
        //
        System.out.println("Temos " + total + " contas abertas!");
    }

    //Declaração de método abstrato, que será implementado pelas classes filhas
    public abstract void deposita(double valor);

    //Declaração de método que efetuará o saque da conta instnaciada. Possui tratamento de exceções e como evoca um método que possui exceção checkd, há necessidade de incluir a declaração de lançamento de exceção
    public void saca(double valor)throws SaldoInsuficienteException{
        //COndicional que testará se existe valor do atributo saldo pertinente para realizar a operação. Caso false, atingirá o tratamento de exceção com exibição de mensagem pertinente 
        if(this.saldo < valor){
            throw new SaldoInsuficienteException("Saque nao ocorreu. Saldo: " + this.saldo + ". Valor a sacar: " + valor);
        }
        //Caso não haja exceção, será decrescentado o conteúdo da variável valor no atributo saldo
        this.saldo -= valor + 0.2;//O valor somado de 0.2 será uma taxa de saque
    }

    //Declaração de método que realizará transferência de valores entre contas, onde a validação de saldos ocorre dentro do método .saca(arg). Método este que possui tratamento de exceção. Como evoca um método que possui exceção checkd, há necessidade de incluir a declaração de lançamento de exceção
    public void transfere(double valor, Conta destino) throws SaldoInsuficienteException{
        //Evocação de método que fará um decremento do conteúdo atribuído à variável valor no saldo do objeto de referência contida na variável destino. O método saca possui tratamento de exceção e em caso de ocorrÊncia, não executará o depósito
        this.saca(valor);
        //Evocação de método que fará um incremento do conteúdo atribuído à variável valor no saldo do objeto de referência contida na variável destino
        destino.deposita(valor);
    }

    //Declaração de método getter devido ao encapsulamento dos atributos
    public double getSaldo(){
        return this.saldo;
    }
    public int getAgencia(){
        return this.agencia;
    }
    public int getNumero(){
        return this.numero;
    }
    public static int getTotal(){
        return Conta.total;
    }

    //Declaração de método setter devido ao encapsulamento dos atributos
    public void setSaldo(double valor){
        this.saldo = valor;
    }
    public void setAgencia(int agencia){
        this.agencia = agencia;
    }
    public void setNumero(int numero){
        this.numero = numero;
    }
}

File: Python/20230826_Entendendo_POO/Conta.py
# Classe que abstrai o conceito de conta bancária com os atributos __numero de iedentificação, titular, saldo atual e limite de crédito"
class Conta:
    # Declaração de construtor com os atributos da classe
    def __init__(self, numero, titular, saldo, limite):
        self.__numero = numero
        self.__titular = titular
        self.__saldo = saldo
        self.__limite = limite
        self.__codigoBanco = "001"

    # Métodos públicos
    # Declaração de método que exibe o titular e o saldo da conta
    def extrato(self):
        print("O saldo da conta {} é R${}.".format(self.__titular, self.__saldo))
    # Declaração de método que fará o incremento de um valor ao atributo saldo
    def deposita(self, valorDeposita):
        self.__saldo += valorDeposita
    # Declaração de método que fará o decréscimo de um valor ao atributo saldo
    def saca(self, valorSaca):
        if(self.__verificaLimite(valorSaca)):
            self.__saldo -= valorSaca
        else:
            print("Valor requisitado maior do que a soma do saldo e do limite!")
    # Declaração de método que fará a transferência de valor entre a conta origem (evocadora do método) e a conta de destino
    def transfere(self, destino, valor):
        self.saca(valor)
        destino.deposita(valor)
    
    # Métodos Privados
    # Método auxiliar que fará a validação se há saldo para realizar o saque
    def __verificaLimite(self, valorSacado):
        return valorSacado <= (self.__saldo + self.__limite)
    
    # Métodos estáticos
    #
    @staticmethod
    def codigoBanco():
        return {'BB': '001', 'Caixa': '104', 'Bradesco':'237'}
            
    #Declaração de métodos getter e setter usando annotations
    # Métodos getter
    @property
    def numero(self):
        return self.__numero
    @property
    def titular(self):
        return self.__titular
    @property
    def saldo(self):
        return self.__saldo
    @property
    def limite(self):
        return self.__limite
    @property
    def codigoBancoConta(self):
        return self.__codigoBanco
    # Métodos setter
    @limite.setter
    def limite(self, valor):
        self.__limite = valor

        
    